File: lib/widgets/password_prompt.dart
import 'package:flutter/material.dart';

import '../components/icon.dart';
import '../theme/theme.dart';

class PasswordForm extends StatefulWidget {
  const PasswordForm({
    super.key,
    required this.message,
    required this.actionLabel,
    required this.onSubmit,
    this.title,
    this.confirm = false,
    this.minLength = 0,
    this.onCancel,
  });

  final String? title;

  final String message;
  final String actionLabel;

  final bool confirm;
  final int minLength;

  /// Returns an error to display, or null on success. Async so the caller can
  /// verify against the vault before the form dismisses.
  final Future<String?> Function(String password) onSubmit;

  final VoidCallback? onCancel;

  @override
  State<PasswordForm> createState() => _PasswordFormState();
}

class _PasswordFormState extends State<PasswordForm> {
  final _password = TextEditingController();
  final _confirm = TextEditingController();
  bool _busy = false;
  String? _error;

  @override
  void dispose() {
    _password.dispose();
    _confirm.dispose();
    super.dispose();
  }

  Future<void> _submit() async {
    if (_busy) return;
    final password = _password.text;

    if (password.isEmpty) {
      setState(() => _error = 'Enter a master password');
      return;
    }
    if (widget.minLength > 0 && password.length < widget.minLength) {
      setState(() => _error = 'Use at least ${widget.minLength} characters');
      return;
    }
    if (widget.confirm && password != _confirm.text) {
      setState(() => _error = 'Passwords do not match');
      return;
    }

    setState(() {
      _busy = true;
      _error = null;
    });
    final problem = await widget.onSubmit(password);
    if (!mounted) return;
    setState(() {
      _busy = false;
      _error = problem;
    });
  }

  @override
  Widget build(BuildContext context) {
    final colors = context.appColors;

    return Column(
      mainAxisSize: MainAxisSize.min,
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        Center(
          child: QIconBadge(
            icon: Icons.lock_outline,
            color: colors.accent,
            size: 60,
            iconSize: 32,
            borderRadius: 16,
          ),
        ),
        const SizedBox(height: 18),
        if (widget.title != null) ...[
          Text(
            widget.title!,
            textAlign: TextAlign.center,
            style: TextStyle(
              color: colors.textPrimary,
              fontSize: 22,
              fontWeight: FontWeight.w700,
            ),
          ),
          const SizedBox(height: 8),
        ],
        Text(
          widget.message,
          textAlign: TextAlign.center,
          style: TextStyle(color: colors.textMuted, fontSize: 13, height: 1.35),
        ),
        const SizedBox(height: 22),
        _Field(
          controller: _password,
          label: 'Master password',
          autofocus: true,
          onSubmitted: widget.confirm ? null : (_) => _submit(),
        ),
        if (widget.confirm) ...[
          const SizedBox(height: 10),
          _Field(
            controller: _confirm,
            label: 'Confirm password',
            onSubmitted: (_) => _submit(),
          ),
        ],
        if (_error != null) ...[
          const SizedBox(height: 12),
          Text(
            _error!,
            textAlign: TextAlign.center,
            style: TextStyle(color: colors.danger, fontSize: 12.5),
          ),
        ],
        const SizedBox(height: 20),
        Row(
          children: [
            if (widget.onCancel != null) ...[
              Expanded(
                child: SizedBox(
                  height: 46,
                  child: TextButton(
                    onPressed: _busy ? null : widget.onCancel,
                    style: TextButton.styleFrom(
                      foregroundColor: colors.textSecondary,
                      shape: RoundedRectangleBorder(
                        borderRadius: BorderRadius.circular(12),
                      ),
                    ),
                    child: const Text('Cancel'),
                  ),
                ),
              ),
              const SizedBox(width: 10),
            ],
            Expanded(
              flex: 2,
              child: SizedBox(
                height: 46,
                child: FilledButton(
                  onPressed: _busy ? null : _submit,
                  style: FilledButton.styleFrom(
                    backgroundColor: colors.accent,
                    foregroundColor: colors.onAccent,
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(12),
                    ),
                  ),
                  child: _busy
                      ? SizedBox(
                          width: 18,
                          height: 18,
                          child: CircularProgressIndicator(
                            strokeWidth: 2,
                            color: colors.onAccent,
                          ),
                        )
                      : Text(widget.actionLabel),
                ),
              ),
            ),
          ],
        ),
      ],
    );
  }
}

class _Field extends StatelessWidget {
  const _Field({
    required this.controller,
    required this.label,
    this.autofocus = false,
    this.onSubmitted,
  });

  final TextEditingController controller;
  final String label;
  final bool autofocus;
  final ValueChanged<String>? onSubmitted;

  @override
  Widget build(BuildContext context) {
    final colors = context.appColors;
    OutlineInputBorder border(Color color, [double width = 1]) =>
        OutlineInputBorder(
          borderRadius: BorderRadius.circular(12),
          borderSide: BorderSide(color: color, width: width),
        );

    return TextField(
      controller: controller,
      autofocus: autofocus,
      obscureText: true,
      onSubmitted: onSubmitted,
      style: TextStyle(color: colors.textPrimary, fontSize: 15),
      decoration: InputDecoration(
        labelText: label,
        labelStyle: TextStyle(color: colors.textMuted, fontSize: 14),
        filled: true,
        fillColor: colors.card,
        border: border(colors.divider),
        enabledBorder: border(colors.divider),
        focusedBorder: border(colors.accent, 1.5),
      ),
    );
  }
}

/// [verify] runs before the dialog closes and returns an error to show inline.
Future<String?> promptForPassword(
  BuildContext context, {
  required String message,
  String actionLabel = 'Confirm',
  bool confirm = false,
  int minLength = 0,
  Future<String?> Function(String password)? verify,
}) {
  final colors = context.appColors;
  return showDialog<String>(
    context: context,
    barrierColor: colors.dialogBarrier,
    builder: (dialogContext) => Dialog(
      backgroundColor: colors.background,
      insetPadding: const EdgeInsets.symmetric(horizontal: 24, vertical: 32),
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
      child: ConstrainedBox(
        constraints: const BoxConstraints(maxWidth: 380),
        child: SingleChildScrollView(
          padding: const EdgeInsets.fromLTRB(24, 28, 24, 20),
          child: PasswordForm(
            message: message,
            actionLabel: actionLabel,
            confirm: confirm,
            minLength: minLength,
            onCancel: () => Navigator.of(dialogContext).pop(),
            onSubmit: (password) async {
              final problem = await verify?.call(password);
              if (problem != null) return problem;
              if (dialogContext.mounted) {
                Navigator.of(dialogContext).pop(password);
              }
              return null;
            },
          ),
        ),
      ),
    ),
  );
}

File: shared/src/commonMain/kotlin/com/example/ohmyssh/widgets/PasswordPrompt.kt
package com.example.ohmyssh.widgets

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import com.example.ohmyssh.components.QIconBadge
import com.example.ohmyssh.theme.appColors
import com.example.ohmyssh.ui.Dialogs
import kotlinx.coroutines.launch

@Composable
fun PasswordForm(
    message: String,
    actionLabel: String,
    onSubmit: suspend (password: String) -> String?,
    title: String? = null,
    confirm: Boolean = false,
    minLength: Int = 0,
    onCancel: (() -> Unit)? = null,
) {
    val colors = appColors
    val scope = rememberCoroutineScope()
    var password by remember { mutableStateOf("") }
    var confirmValue by remember { mutableStateOf("") }
    var busy by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    fun submit() {
        if (busy) return
        if (password.isEmpty()) {
            error = "Enter a master password"
            return
        }
        if (minLength > 0 && password.length < minLength) {
            error = "Use at least $minLength characters"
            return
        }
        if (confirm && password != confirmValue) {
            error = "Passwords do not match"
            return
        }
        busy = true
        error = null
        scope.launch {
            val problem = onSubmit(password)
            busy = false
            error = problem
        }
    }

    Column(Modifier.fillMaxWidth()) {
        Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
            QIconBadge(
                icon = Icons.Outlined.Lock,
                color = colors.accent,
                size = 60.dp,
                iconSize = 32.dp,
                borderRadius = 16.dp,
            )
        }
        Spacer(Modifier.height(18.dp))
        if (title != null) {
            Text(
                title,
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth(),
                style = TextStyle(
                    color = colors.textPrimary,
                    fontSize = 22.sp,
                    fontWeight = FontWeight.W700,
                ),
            )
            Spacer(Modifier.height(8.dp))
        }
        Text(
            message,
            textAlign = TextAlign.Center,
            modifier = Modifier.fillMaxWidth(),
            style = TextStyle(color = colors.textMuted, fontSize = 13.sp, lineHeight = 17.5.sp),
        )
        Spacer(Modifier.height(22.dp))
        QTextField(
            value = password,
            onValueChange = { password = it },
            label = "Master password",
            obscure = true,
            autofocus = true,
            onSubmitted = { if (!confirm) submit() },
        )
        if (confirm) {
            Spacer(Modifier.height(10.dp))
            QTextField(
                value = confirmValue,
                onValueChange = { confirmValue = it },
                label = "Confirm password",
                obscure = true,
                onSubmitted = { submit() },
            )
        }
        val problem = error
        if (problem != null) {
            Spacer(Modifier.height(12.dp))
            Text(
                problem,
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth(),
                style = TextStyle(color = colors.danger, fontSize = 12.5.sp),
            )
        }
        Spacer(Modifier.height(20.dp))
        Row(Modifier.fillMaxWidth()) {
            if (onCancel != null) {
                TextButton(
                    onClick = { if (!busy) onCancel() },
                    enabled = !busy,
                    shape = RoundedCornerShape(12.dp),
                    colors = ButtonDefaults.textButtonColors(contentColor = colors.textSecondary),
                    modifier = Modifier.weight(1f).height(46.dp),
                ) { Text("Cancel") }
                Spacer(Modifier.width(10.dp))
            }
            Button(
                onClick = { submit() },
                enabled = !busy,
                shape = RoundedCornerShape(12.dp),
                colors = ButtonDefaults.buttonColors(
                    containerColor = colors.accent,
                    contentColor = colors.onAccent,
                    disabledContainerColor = colors.accent.copy(alpha = 0.6f),
                    disabledContentColor = colors.onAccent,
                ),
                modifier = Modifier.weight(2f).height(46.dp),
            ) {
                if (busy) {
                    CircularProgressIndicator(
                        strokeWidth = 2.dp,
                        color = colors.onAccent,
                        modifier = Modifier.size(18.dp),
                    )
                } else {
                    Text(actionLabel)
                }
            }
        }
    }
}

suspend fun promptForPassword(
    message: String,
    actionLabel: String = "Confirm",
    confirm: Boolean = false,
    minLength: Int = 0,
    verify: (suspend (password: String) -> String?)? = null,
): String? = Dialogs.show { dismiss ->
    val colors = appColors
    Dialog(onDismissRequest = { dismiss(null) }) {
        Surface(
            shape = RoundedCornerShape(16.dp),
            color = colors.background,
            modifier = Modifier.widthIn(max = 380.dp),
        ) {
            Column(
                Modifier
                    .verticalScroll(rememberScrollState())
                    .padding(start = 24.dp, top = 28.dp, end = 24.dp, bottom = 20.dp),
            ) {
                PasswordForm(
                    message = message,
                    actionLabel = actionLabel,
                    confirm = confirm,
                    minLength = minLength,
                    onCancel = { dismiss(null) },
                    onSubmit = { password ->
                        val problem = verify?.invoke(password)
                        if (problem != null) {
                            problem
                        } else {
                            dismiss(password)
                            null
                        }
                    },
                )
            }
        }
    }
}
